File: easy/array/python/1-two-sum.py
import enum
from typing import List
from collections import defaultdict


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_dict = defaultdict(list)
        # for idx, n in enumerate(nums):
        #     nums_dict[n].append(idx)

        # for idx, n in enumerate(nums):
        #     if target - n in nums_dict:
        #         if target - n == n:
        #             if len(nums_dict[target - n]) == 1:
        #                 continue
        #             else:
        #                 return (idx, nums_dict[target - n][1])

        #         return (idx, nums_dict[target - n][0])
        ##
        for idx, n in enumerate(nums):
            if target - n in nums_dict:
                return (idx, nums_dict[target - n][0])
            else:
                nums_dict[n].append(idx)

File: easy/array/cpp/1-two-sum.cpp
#include "../../../stdc++.h"

using std::unordered_map;
using std::vector;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> map;
        for (int i = 0; i < nums.size(); i++)
        {
            int value = nums[i];
            if (map.find(target - value) != map.end())
            {
                int otherindex = map.find(target - value)->second;
                return {i, otherindex};
            }
            else
            {
                map[value] = i;
            }
        }
        return {};
    }
};

int main()
{
    Solution a;
    vector<int> vect = {1, 2, 3, 4};
    a.twoSum(vect, 5);
}
